src/net_processing.cpp

```diff
@@ -1647,8 +1663,8 @@ static void RelayAddress(const CNode& originator,
     std::array<std::pair<uint64_t, CNode*>,2> best{{{0, nullptr}, {0, nullptr}}};
     assert(nRelayNodes <= best.size());
 
-    auto sortfunc = [&best, &hasher, nRelayNodes, &originator, &addr](CNode* pnode) {
-        if (pnode->RelayAddrsWithConn() && pnode != &originator && pnode->IsAddrCompatible(addr)) {
+    auto sortfunc = [&best, &hasher, nRelayNodes, &originator, &addr, peerman = this](CNode* pnode) {
+        if (peerman->IsAddrRelayPeer(*pnode) && pnode != &originator && pnode->IsAddrCompatible(addr)) {
```

Comment:
Is it possible to avoid this extreme coupling between the net and net_processing layers? Here, net_processing is constructing a callback (`sortfunc`) which it hands to net via `ForEachNodeThen()`. In that callback, net calls into a net_processing function (`IsAddRelayPeer()`), which in turn calls into a function in the net layer (`RelayAddrsWithConn()`). This net_processing -> net -> net_processing -> net layering makes it very difficult to construct a mental model of which component is responsible for what functions, and makes future work to rationalize the interface much more difficult.

Reply:
Haven't looked at the details, but I'd presume that `RelayAddrsWithConn` will be (can be) removed once the addr filter is moved to net processing and initialized only when needed.

Reply:
Yeah I think this will get better once we finish moving things from `net` to `net_processing`.  In this case, I've added a new variable and member function to `net_processing` (which is where I think it should live in the long run) but until the rest of it moves, we're going to have to look in both places for some logic.

Reply:
I don't agree with this line of logic. I think it'd be far better to remove complexity first and then implement this cleanly, rather than add on more complexity and hoping to remove it later. Also note that this change is a pessimization: it locks and releases `m_peer_mutex` for every peer (up to 125 times). My understanding is that the `ForEachNode()` helpers exist in part to avoid this kind of repeated locking/unlocking.

Reply:
@jnewbery You've been doing the most work on moving things out of `net` -- can you be more specific about what you want to see in this implementation?  I was trying to follow a guideline of "don't add new things to `net` if they should end up in `net_processing`", in the hopes of not making that work any harder.  Are you suggesting that I add this to `net` to make this implementation cleaner, or wait for more of your refactoring work to land, or something else?

Reply:
> Are you suggesting that I add this to net to make this implementation cleaner, or wait for more of your refactoring work to land, or something else?

Yes, certainly rebasing this on the https://github.com/jnewbery/bitcoin/tree/2020-06-cs-main-split branch from https://github.com/bitcoin/bitcoin/issues/19398 would remove this tight net-net_processing coupling. We all agree that we want addr relay logic to live entirely in net_processing. My point here is that changes we make now should be done in a way that moves in that direction, not in a way that makes it more difficult.

Reply:
Thanks for being clear about your preference here.  I don't think it's worth gating this work on all those commits landing in master, but if other reviewers prefer to see that work land first, then of course I'll rebase after your commits are merged.

For now I've gone ahead and simplified this PR by just putting `m_disable_tx` in `CNode`, which makes things no worse than they are today as far as the separation between `net` and `net_processing` (and indeed it simplifies this PR by not adding the complexity you pointed out here in `RelayAddress`).